File: app/src/main/java/com/ryanrvldo/moviecatalogue/data/Repository.java
package com.ryanrvldo.moviecatalogue.data;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.ryanrvldo.moviecatalogue.data.source.local.LocalRepository;
import com.ryanrvldo.moviecatalogue.data.source.model.Movie;
import com.ryanrvldo.moviecatalogue.data.source.model.Search;
import com.ryanrvldo.moviecatalogue.data.source.model.Season;
import com.ryanrvldo.moviecatalogue.data.source.model.TvShow;
import com.ryanrvldo.moviecatalogue.data.source.remote.LoadDataCallback;
import com.ryanrvldo.moviecatalogue.data.source.remote.RemoteRepository;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.MovieResponse;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.TvShowResponse;

import java.util.List;

public class Repository implements DataSource {

    private volatile static Repository mInstance = null;
    private final RemoteRepository remoteRepository;
    private final LocalRepository localRepository;

    private Repository(@NonNull RemoteRepository remoteRepository,
                       @NonNull LocalRepository localRepository) {
        this.remoteRepository = remoteRepository;
        this.localRepository = localRepository;
    }

    public static Repository getInstance(RemoteRepository remoteData, LocalRepository localData) {
        if (mInstance == null) {
            synchronized (Repository.class) {
                if (mInstance == null) {
                    mInstance = new Repository(remoteData, localData);
                }
            }
        }
        return mInstance;
    }

    @Override
    public LiveData<MovieResponse> getMovies(String category) {
        MutableLiveData<MovieResponse> response = new MutableLiveData<>();
        remoteRepository.getMovies(category, new LoadDataCallback<MovieResponse>() {
            @Override
            public void onDataReceived(MovieResponse data) {
                response.postValue(data);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("GetMovies", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<MovieResponse> searchMovies(String query) {
        MutableLiveData<MovieResponse> response = new MutableLiveData<>();
        remoteRepository.searchMovies(query, new LoadDataCallback<MovieResponse>() {
            @Override
            public void onDataReceived(MovieResponse movieResponse) {
                response.postValue(movieResponse);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("SearchMovies", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<MovieResponse> getNewReleaseMovies() {
        MutableLiveData<MovieResponse> response = new MutableLiveData<>();
        remoteRepository.getNewReleaseMovies(new LoadDataCallback<MovieResponse>() {
            @Override
            public void onDataReceived(MovieResponse movieResponse) {
                response.postValue(movieResponse);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("NewReleaseMovies", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<Movie> getMovieDetail(int movieId) {
        MutableLiveData<Movie> response = new MutableLiveData<>();
        remoteRepository.getMovieDetail(movieId, new LoadDataCallback<Movie>() {
            @Override
            public void onDataReceived(Movie movie) {
                response.postValue(movie);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("GetMovieDetail", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<TvShowResponse> getTvShows(String category) {
        MutableLiveData<TvShowResponse> response = new MutableLiveData<>();
        remoteRepository.getTvShows(category, new LoadDataCallback<TvShowResponse>() {
            @Override
            public void onDataReceived(TvShowResponse tvShowResponse) {
                response.postValue(tvShowResponse);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("GetTvShows", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<TvShowResponse> searchTvShows(String query) {
        MutableLiveData<TvShowResponse> response = new MutableLiveData<>();
        remoteRepository.searchTvShows(query, new LoadDataCallback<TvShowResponse>() {
            @Override
            public void onDataReceived(TvShowResponse tvShowResponse) {
                response.postValue(tvShowResponse);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("SearchTvShows", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<TvShowResponse> getNewReleaseTvShows() {
        MutableLiveData<TvShowResponse> response = new MutableLiveData<>();
        remoteRepository.getNewReleaseTvShows(new LoadDataCallback<TvShowResponse>() {
            @Override
            public void onDataReceived(TvShowResponse tvShowResponse) {
                response.postValue(tvShowResponse);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("NewReleaseTvShows", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<TvShow> getTvDetail(int tvShowId) {
        MutableLiveData<TvShow> response = new MutableLiveData<>();
        remoteRepository.getTvDetail(tvShowId, new LoadDataCallback<TvShow>() {
            @Override
            public void onDataReceived(TvShow tvShow) {
                response.postValue(tvShow);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("GetTvDetail", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<Season> getSeasonDetail(int tvId, int seasonNumber) {
        MutableLiveData<Season> response = new MutableLiveData<>();
        remoteRepository.getSeasonDetail(tvId, seasonNumber, new LoadDataCallback<Season>() {
            @Override
            public void onDataReceived(Season season) {
                response.postValue(season);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d("GetSeasons", "onDataNotAvailable");
            }
        });
        return response;
    }

    @Override
    public LiveData<List<Movie>> getFavMovies() {
        return localRepository.getAllFavMovies();
    }

    @Override
    public void insertFavMovie(Movie movie) {
        localRepository.insertMovie(movie);
    }

    @Override
    public void deleteFavMovie(Movie movie) {
        localRepository.deleteMovie(movie);
    }

    @Override
    public Movie getFavMovie(int movieId) {
        return localRepository.getMovie(movieId);
    }

    @Override
    public LiveData<List<TvShow>> getFavTvShows() {
        return localRepository.getAllFavTv();
    }

    @Override
    public void insertFavTv(TvShow tvShow) {
        localRepository.insertTv(tvShow);
    }

    @Override
    public void deleteFavTv(TvShow tvShow) {
        localRepository.deleteTv(tvShow);
    }

    @Override
    public TvShow getFavTv(int tvShowId) {
        return localRepository.getTvShow(tvShowId);
    }

    @Override
    public LiveData<List<Search>> getSearchHistory() {
        return localRepository.getAllSearch();
    }

    @Override
    public void insertSearch(Search search) {
        localRepository.insertSearch(search);
    }

    @Override
    public void deleteSearch(Search search) {
        localRepository.deleteSearch(search);
    }

    @Override
    public void deleteAllSearchHistory() {
        localRepository.deleteAllSearch();
    }

    @Override
    public String getSearch(String query) {
        return localRepository.getSearch(query);
    }
}

File: app/src/main/kotlin/com/ryanrvldo/moviecatalogue/data/Repository.kt
package com.ryanrvldo.moviecatalogue.data

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.ryanrvldo.moviecatalogue.data.local.LocalRepository
import com.ryanrvldo.moviecatalogue.data.model.Movie
import com.ryanrvldo.moviecatalogue.data.model.Search
import com.ryanrvldo.moviecatalogue.data.model.Season
import com.ryanrvldo.moviecatalogue.data.model.TvShow
import com.ryanrvldo.moviecatalogue.data.remote.LoadDataCallback
import com.ryanrvldo.moviecatalogue.data.remote.RemoteRepository
import com.ryanrvldo.moviecatalogue.data.remote.response.MovieResponse
import com.ryanrvldo.moviecatalogue.data.remote.response.TvShowResponse
import com.ryanrvldo.moviecatalogue.data.vo.Resource
import com.ryanrvldo.moviecatalogue.utils.performGetOperation
import timber.log.Timber
import javax.inject.Inject

class Repository @Inject constructor(
    private val remoteRepository: RemoteRepository,
    private val localRepository: LocalRepository,
) : DataSource {

    override fun getMovies(category: String) = performGetOperation {
        remoteRepository.getMovies(category)
    }

    override fun getMovieDetails(movieId: Int): LiveData<Resource<Movie>> = performGetOperation {
        remoteRepository.getMovieDetails(movieId)
    }

    override fun searchMovies(query: String): LiveData<Resource<MovieResponse>> =
        performGetOperation {
            remoteRepository.searchMovies(query)
        }

    override fun getNewReleaseMovies(): LiveData<MovieResponse> {
        val response = MutableLiveData<MovieResponse>()
        remoteRepository.getNewReleaseMovies(object : LoadDataCallback<MovieResponse> {
            override fun onDataReceived(data: MovieResponse) {
                response.postValue(data)
            }

            override fun onDataNotAvailable() {
                Timber.d("onDataNotAvailable")
            }
        })
        return response
    }

    override fun getTvShows(category: String): LiveData<Resource<TvShowResponse>> =
        performGetOperation { remoteRepository.getTvShows(category) }

    override fun getTvShowDetails(tvShowId: Int): LiveData<Resource<TvShow>> = performGetOperation {
        remoteRepository.getTvDetail(tvShowId)
    }

    override fun searchTvShows(query: String): LiveData<Resource<TvShowResponse>> =
        performGetOperation { remoteRepository.searchTvShows(query) }

    override fun getNewReleaseTvShows(): LiveData<TvShowResponse> {
        val response = MutableLiveData<TvShowResponse>()
        remoteRepository.getNewReleaseTvShows(object : LoadDataCallback<TvShowResponse> {
            override fun onDataReceived(data: TvShowResponse) {
                response.postValue(data)
            }

            override fun onDataNotAvailable() {
                Timber.tag("NewReleaseTvShows").d("onDataNotAvailable")
            }
        })
        return response
    }

    override fun getSeasonDetails(tvId: Int, seasonNumber: Int): LiveData<Season> {
        val response = MutableLiveData<Season>()
        remoteRepository.getSeasonDetail(tvId, seasonNumber, object : LoadDataCallback<Season> {
            override fun onDataReceived(data: Season) {
                response.postValue(data)
            }

            override fun onDataNotAvailable() {
                Timber.d("onDataNotAvailable")
            }
        })
        return response
    }

    override fun getFavoriteMovies(): LiveData<List<Movie>> = localRepository.getFavoriteMovies()

    override suspend fun addFavoriteMovie(movie: Movie) = localRepository.addMovie(movie)

    override suspend fun removeFavoriteMovie(movie: Movie) = localRepository.removeMovie(movie)

    override suspend fun isFavoriteMovie(movieId: Int): Boolean =
        localRepository.isMovieExists(movieId)

    override fun getFavoriteTvShows(): LiveData<List<TvShow>> = localRepository.getFavoriteTvShows()

    override suspend fun addFavoriteTvShow(tvShow: TvShow) = localRepository.addTvShow(tvShow)

    override suspend fun removeFavoriteTvShow(tvShow: TvShow) = localRepository.removeTvShow(tvShow)

    override suspend fun isFavoriteTvShow(tvShowId: Int): Boolean =
        localRepository.isTvShowExists(tvShowId)

    override fun getSearchHistories(): LiveData<List<Search>> = localRepository.getSearchHistories()

    override suspend fun addSearchQuery(search: Search) = localRepository.addSearchQuery(search)

    override suspend fun removeSearchQuery(search: Search) =
        localRepository.removeSearchQuery(search)

    override suspend fun removeSearchHistories() = localRepository.removeSearchHistories()

    override suspend fun selectSearchQuery(query: String): String? =
        localRepository.getSearch(query)
}
